File: Algorithm_203_Remove Linked List Elements.cpp
#include <iostream>
using namespace std;
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
ListNode* removeElements(ListNode* head, int data) {
    if (head==NULL) return NULL;
    while (head!=NULL && head->val==data)
    {
        ListNode* deleted = head;
        head = head->next;
        delete deleted;
    }
    ListNode *tmp,*deleted;
    tmp=head;
    while (tmp!=NULL && tmp->next!=NULL)
    {
        if (tmp->next->val==data)
        {
            deleted=tmp->next;
            tmp->next=tmp->next->next;
            delete deleted;
        }
        else
        tmp=tmp->next;
    }
    return head;
}
int main()
{
	ListNode *n1=new ListNode(1);
	ListNode *n2=new ListNode(2);
	ListNode *n3=new ListNode(3);
	n1->next=n2;
	n2->next=n3;
	n1=removeElements(n1,3);
	ListNode *tmp=n1;
	while (tmp!=NULL)
	{
		 cout<<tmp->val<<' ';
		 tmp=tmp->next;
	}
}

File: Algorithm_203_Remove Linked List Elements.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

// 方法：构建一个新的头结点dummyHead，直接遍历链表，pre记录cur的上一个节点

class Solution {
    public ListNode removeElements(ListNode head, int val) {
        ListNode dummyHead = new ListNode(0, head);
        ListNode pre = dummyHead, cur = head;
        while (cur != null) {
            ListNode next = cur.next;
            if (cur.val == val) {
                pre.next = next;
                cur = next;
            } else {
                pre = cur;
                cur = next;
            }
        }
        return dummyHead.next;
    }
}